content/en/blog/2023/exponential-histograms.md

```diff
@@ -0,0 +1,196 @@
+negative scale, as the scale grows, the base shrinks and vice versa. As will be
+shown later, this is the fundamental reason that a greater scale factor results
+in smaller buckets and a higher resolution histogram.
+
+## Bucket calculation
+
+Given a scale factor and its resulting base, we can calculate every possible
+bucket in the histogram. From the base, the upper bound of each bucket at index
+`i` is defined to be `base ^ (i + 1)`, with the first bucket lower boundary
+of 1. Because of this, the upper boundary of the first bucket at index 0 is also
+exactly the base. For now, we will only consider nonnegative indices, but
+negative indexed buckets are also possible and define all buckets between 0
+and 1. Keeping with our example using a scale of 3 and resulting base of
+1.090508, the third bucket at index 2 has an upper bound of
+`1.090508^(2+1) = 1.29684`. The following table shows upper bounds for the first
+10 buckets of a few different scale factors:
+
+| index | scale -1 | scale 0 | scale 1 | scale 3 |
+| ----- | -------- | ------- | ------- | ------- |
+| -1    | **1**    | **1**   | **1**   | **1**   |
+| 0     | **4**    | 2       | 1.4142  | 1.0905  |
+| 1     | **16**   | **4**   | 2       | 1.1892  |
+| 2     | 64       | 8       | 2.8284  | 1.2968  |
+| 3     | 256      | **16**  | **4**   | 1.4142  |
+| 4     | 1024     | 32      | 5.6569  | 1.5422  |
+| 5     | 4096     | 64      | 8       | 1.6818  |
+| 6     | 16384    | 128     | 11.3137 | 1.8340  |
+| 7     | 65536    | 256     | **16**  | 2       |
+| 8     | 262144   | 512     | 22.6274 | 2.1810  |
+| 9     | 1048576  | 1024    | 32      | 2.3784  |
+
+I've bolded some of the values here to show an important property of exponential
+histograms called _perfect subsetting_.
+
+## Perfect subsetting
+
+In the chart above, some of the bucket boundaries are shared between histograms
+with differing scale factors. In fact, each time the scale factor increases by
+1, exactly 1 boundary is inserted between each existing boundary. This feature
+is called perfect subsetting because each set of boundaries for a given scale
+factor is a perfect subset of the boundaries for any histogram with a greater
+scale factor.
+
+Because of this, histograms with differing scale factors can be normalized to
+whichever has the lesser scale factor by combining neighboring buckets. This
+means that histograms with different scale factors can still be combined into a
+single histogram with exactly the precision of the least precise histogram being
+combined. For example, histogram _A_ with scale 3 and histogram _B_ with scale 2
+can be combined into a single histogram _C_ with scale 2 by first summing each
+pair of neighboring buckets in _A_ to form histogram _A'_ with scale 2. Then,
+each bucket in _A'_ is summed with the corresponding bucket of the same index in
+_B_ to make _C_.
+
+## Error rate
+
+The error rate of a histogram is defined as the average relative error when
```

Comment:
The error of a histogram is defined as the maximum relative error when...

Reply:
Isn't it possible to have an absolute error of the entire bucket width?

If a bucket goes from 1 to 2, and you have 100 values of 1.0001, using linear interpolation to select the 99th point in the bucket would result in 1.99 when the actual value is 1.0001. All of the calculations used in the OTEP and spec use the bucket midpoint for the error calculation. To me, that indicates it is the average or expected relative error, but not the maximum.